File: Builder/Builder01.js
//Definido parte de um carro
class Motor{
    constructor(tipo){
        this.tipo = tipo;
    }
}

class Carroceria{
    constructor(estilo){
        this.estilo = estilo;
    }
}

class Rodas{
    constructor(tamanho){
        this.tamanho = tamanho;
    }
}

class Interior{
    constructor(cor){
        this.cor = cor;
    }
}

//Builder
class CarroBuilder{
    constructor(){
        this.motor = null;
        this.carroceria = null;
        this.rodas = null;
        this.interior = null;
    }

    addMotor(tipo){
        this.motor = new Motor(tipo);
        return this;
    }

    addCarroceria(estilo){
        this.carroceria = new Carroceria(estilo);
        return this;
    }

    addRodas(tamanho){
        this.rodas =  new Rodas(tamanho);
        return this;
    }

    addInterior(cor){
        this.interior = new Interior(cor);
        return this;
    }

    construir(){
        return new Carro(this.motor, this.carroceria, this.rodas, this.interior);
    }
}

//Construir carro
class Carro{
    constructor(motor, carroceria, rodas, interior){
        this.motor = motor;
        this.carroceria = carroceria;
        this.rodas = rodas;
        this.interior = interior;
    }

    mostrarDetalhes(){
        console.log(`Carro com motor: ${this.motor.tipo}\n carroceria: ${this.carroceria.estilo}\n rodas: ${this.rodas.tamanho}\n interior: ${this.interior.cor}\n`);
    }
}

//Uso do modelo builder - construindo um carro
const builder = new CarroBuilder();

const carroEsportivo = builder
    .addMotor('V8')
    .addCarroceria('Esportivo')
    .addRodas(18)
    .addInterior("Vermelho")
    .construir()

const carroSedan = builder
    .addMotor('1.4 LT')
    .addCarroceria('Sedan')
    .addRodas(15)
    .addInterior("Preto")
    .construir()
    
carroEsportivo.mostrarDetalhes();
carroSedan.mostrarDetalhes();

File: Builder/Builder01.py
#Definido parte de um carro
class Motor:
    def __init__(self,tipo):
        self.tipo = tipo

class Carroceria:
    def __init__(self,estilo):
        self.estilo = estilo

class Rodas:
    def __init__(self, tamanho):
        self.tamanho = tamanho

class Interior:
    def __init__(self,cor):
        self.cor = cor


#Builder
def __init__(self):
    self.reset()

class CarroBuilder:
    def __init__(self):
        self.motor = None
        self.carroceria = None
        self.rodas = None
        self.interior = None

    def addMotor(self, tipo):
        self.motor = Motor(tipo)
        return self
    
    def addCarroceria(self, estilo):
        self.carroceria = Carroceria(estilo)
        return self
    
    def addRodas(self, tamanho):
        self.rodas = Rodas(tamanho)
        return self
    
    def addInterior(self, cor):
        self.interior = Interior(cor)
        return self
    
    def construir(self):
        carro = Carro(self.motor, self.carroceria, self.rodas, self.interior)
        return carro
    
#Construir carro    
class Carro:
    def __init__(self, motor, carroceria, rodas, interior):
        self.motor = motor
        self.carroceria = carroceria
        self.rodas = rodas
        self.interior = interior

    def mostrarDetalhes(self):
        print(f"Carro com motor: {self.motor.tipo}\n carroceria: {self.carroceria.estilo}\n rodas: {self.rodas.tamanho}\n interior: {self.interior.cor}\n")

#Uso do modelo builder - construindo um carro
carroEsportivo = (
    CarroBuilder()
    .addMotor('V8')
    .addCarroceria('Esportivo')
    .addRodas(18)
    .addInterior('Vermelho')
    .construir()
)

carroSedan = (
    CarroBuilder()
    .addMotor('1.4 LT')
    .addCarroceria('Sedan')
    .addRodas(14)
    .addInterior('Preto')
    .construir()
)

carroEsportivo.mostrarDetalhes()
carroSedan.mostrarDetalhes()

